### PR TITLE
services: Add dropbox basic support

### PR DESCRIPTION
**Pull request by ccamel**
Hi @Xuanwo!

Since I needed to integrate with Dropbox, I've implemented a (basic) `storager` for this purpose.
It may be of interest for your project (a very promising one!).

Sadly it comes with no unit-tests, but the tests I have been able to perform so far have succeeded.

Any comments/suggestions are welcome.

### CODE REVIEW

File: services/dropbox/storager.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+					Type:     types.ObjectTypeFile,
+					Name:     filepath.Join(path, info.Name),
+					Metadata: make(metadata.Metadata),
+				}
+				o.SetId(info.Id)
+
+				if opt.HasDirFunc {
+					opt.DirFunc(o)
+				}
+			}
```
Comment (Xuanwo) [inline services/dropbox/storager.go]: Add `default` case for type switch failed.

File: services/dropbox/storager.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+		}
+
+		if result.HasMore {
```
Comment (Xuanwo) [inline services/dropbox/storager.go]: Use the style following looks more clear for me.

```go
if !result.HasMore {
	break
}

result, err = s.service.ListFolderContinue(&files.ListFolderContinueArg{
	Cursor: result.Cursor,
})
if err != nil {
	return fmt.Errorf(errorMessage, s, path, err)
}
```

File: services/dropbox/storager.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	"github.com/Xuanwo/storage/types/metadata"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/dropbox"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/dropbox/files"
+)
+
+// Storage is the dropbox client.
+//
+//go:generate ../../internal/bin/meta
+type Storage struct {
+	service files.Client
```
Comment (Xuanwo) [inline services/dropbox/storager.go]: `service` has special meaning for `Servicer`, let's use `client` here.

File: services/dropbox/storager.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+			Size:      int64(meta.Size),
+			UpdatedAt: meta.ServerModified,
+			Metadata:  make(metadata.Metadata),
+		}
+		o.SetChecksum(meta.ContentHash)
+
+		return o, nil
+	case *files.FolderMetadata:
+		o := &types.Object{
+			Type: types.ObjectTypeFile,
```
Comment (Xuanwo) [inline services/dropbox/storager.go]: `types.ObjectTypeDir` insdead for folder.

**Comment by Xuanwo** [comment]
Let's move id to `Object` struct instead of metadata. https://github.com/Xuanwo/storage/pull/55

File: coreutils/open.go [hunk 5]
```diff
@@ -132,6 +135,27 @@ func Open(cfg string) (srv storage.Servicer, store storage.Storager, err error)
...
+	case dropbox.Type:
+		store, err = dropbox.New()
+		if err != nil {
+			err = fmt.Errorf(errorMessage, cfg, err)
+			return
+		}
+
+		args := []*types.Pair{pairs.WithWorkDir(namespace)}
+		for _, v := range opt {
+			if v.Key == pairs.Credential {
```
[the remark this answers is not in the document]
Comment (ccamel) [reply coreutils/open.go]: Oh nice. Will fix that.

File: services/dropbox/storager.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	return fmt.Sprintf(
+		"Storager dropbox {WorkDir: %s}",
+		"/"+s.workDir,
+	)
+}
+
+// Metadata implements Storager.Metadata
+func (s *Storage) Metadata() (m metadata.Storage, err error) {
+	m = metadata.Storage{
+		Name:     "",
```
[the remark this answers is not in the document]
Comment (ccamel) [reply services/dropbox/storager.go]: What this field is intended for? For `s3` for instance, the name refers to the bucketName, but for dropbox you want to get the name of the account?

[the remark this answers is not in the document]
Comment (Xuanwo) [reply services/dropbox/storager.go]: Yes, can we do that？

**Comment by Xuanwo** [comment]
@ccamel I'm trying to propose a new draft to address this issue, maybe we can ignore this meta or add a FIXME so that we can merge this PR?

**Comment by Xuanwo** [comment]
Maybe we can store this value in `ETag`:

> Entity tag uniquely representing the requested resource. They are a string of ASCII characters placed between double quotes, like "675af34563dc-tr34". The method by which ETag values are generated is not specified. Often, a hash of the content, a hash of the last modification timestamp, or just a revision number is used. For example, MDN uses a hexadecimal hash of the wiki article content.

ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

So the different between `content-md5` and `etag` is:

- `content-md5` can be used to compare between different services.
- `etag` allow every service has their own implement and only comparable within the same service.

It's the implementer's duty to decide which one (or both) to store.

Any idea to that?

File: services/dropbox/storager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ccamel) [reply services/dropbox/storager.go]: No I don't think so.

[the remark this answers is not in the document]
Comment (Xuanwo) [reply services/dropbox/storager.go]: OK, let's keep it empty.

**Pull request merged**